### PR TITLE
Add `Aws_efs` module and `secotrec-aws-efs` app

### PR DESCRIPTION
**Pull request by smondet**
The app can setup AWS-EFS file-systems on AWS nodes created with
`secotrec-aws-node`, all operations are nicely idempotent:

    $ secotrec-aws-efs up
        --name smondet-efs --ssh ec2-user@ec2-example.compute-42.amazonaws.com

    SECOTREC: [EFS:smondet-efs] Checking file-system-id
    SECOTREC: [EFS:smondet-efs] Building file-system-id
    SECOTREC: [EFS:smondet-efs]  -> file-system-id was just created: fs-3945ee70
    SECOTREC: [EFS:smondet-efs] Using: Subnet: subnet-8f7df1f8, Secgrp: sg-1f2de660
    SECOTREC: [EFS:smondet-efs] Checking mount-target-id
    SECOTREC: [EFS:smondet-efs] Building mount-target-id
    SECOTREC: [EFS:smondet-efs]  -> mount-target-id was just created: fsmt-bf47faf6
    SECOTREC: [EFS:smondet-efs] Authorizing :2049 traffic within group sg-1f2de660
    SECOTREC: [EFS:smondet-efs]  -> NFSv4 traffic was already authorized
    SECOTREC: [EFS:smondet-efs] Waiting for mount-target to be really available.
    0.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.17.18.19.20.21.22.23.24.25.26.
    SECOTREC: [EFS:smondet-efs] Mounting 172.31.45.55:/ at /mnt-smondet-efs

    $ secotrec-aws-efs describe
        --name smondet-efs --ssh ec2-user@ec2-example.compute-42.amazonaws.com

    SECOTREC: [EFS:smondet-efs] File-system-id: fs-3945ee70:
    ``````````json
    {
        "FileSystems": [
            {
                "SizeInBytes": {
                    "Value": 6144
                },
                "CreationToken": "smondet-efs",
                "CreationTime": 1489613265.0,
                "PerformanceMode": "generalPurpose",
                "FileSystemId": "fs-3945ee70",
                "NumberOfMountTargets": 1,
                "LifeCycleState": "available",
                "OwnerId": "020590027510"
            }
        ]
    }

    ``````````
    SECOTREC: [EFS:smondet-efs] Mount-Targets:
    ``````````
    MOUNTTARGETS	fs-3945ee70	172.31.45.55	available	fsmt-bf47faf6	eni-762f1f95	020590027510	subnet-8f7df1f8

    ``````````
    SECOTREC: [EFS:smondet-efs] Local-mount:
    ``````````
    172.31.45.55:/ on /mnt-smondet-efs type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=172.31.36.66,local_lock=none,addr=172.31.45.55)

    ``````````
    SECOTREC: [EFS:smondet-efs] Done.

    $ secotrec-aws-efs down
        --name smondet-efs --ssh ec2-user@ec2-example.compute-42.amazonaws.com

    SECOTREC: [EFS:smondet-efs] File-system-ID: fs-3945ee70, Mount-Target-ID: fsmt-bf47faf6
    SECOTREC: [EFS:smondet-efs] Unmounting `/mnt-smondet-efs`...
    ``````````

    ``````````
    SECOTREC: Deleting mount-target.
    SECOTREC: Waiting for the FS to not be “in use”: `1` user(s) now...
    0.1.
    SECOTREC: EFS-smondet-efs: Deleting file-system.

### CODE REVIEW

File: src/app/aws_efs.ml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  (** Name of the AWS-EFS file-system (a.k.a. "creation-token"). *)
+
+  ssh_aws_node: string;
+  (** How to connect to the AWS node over SSH (e.g. "ec2-user@instance.example.com"). *)
+
+} [@@deriving cmdliner]
+
+
+let run_genspio_over_ssh efs_test g =
+  Secotrec.Common.write_file "/tmp/gsp.sh" ~content:(Genspio.Language.to_many_lines g);
```
Comment (ihodes) [inline src/app/aws_efs.ml]: Maybe a true random tmp file name? This seems like an accident waiting to happen?

File: src/lib/aws_efs.ml [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+end
+
+let tr_remove_new_lines = Genspio.EDSL.exec ["tr"; "-d"; "\\n"]
+
+type guess_value = [ `From_metadata | `Value of string ] [@@deriving yojson]
+
+type t = {
+  name: string [@main];
+  guess_subnet: guess_value [@default `From_metadata];
+  guess_secgroup: guess_value [@default `From_metadata];
```
Comment (ihodes) [inline src/lib/aws_efs.ml:35]: Why `guess`? Is it a guess when the Value constructor is passed?

File: src/lib/aws_efs.ml [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        string "--file-system-id"; fs_id;
+        string "--subnet-id"; subnet_id;
+        string "--security-groups"; secgrp_id;
+        string "--query"; string "MountTargetId";
+      ] in
+    aws_get_or_create t "mount-target-id" ~get ~create
+
+  let curl_metadata_item path =
+    let uri =
+      string_concat [string "http://169.254.169.254/latest/meta-data/"; path] in
```
Comment (ihodes) [inline src/lib/aws_efs.ml]: Could be helpful to comment/name this URL

File: src/lib/aws_efs.ml [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+type t = {
+  name: string [@main];
+  guess_subnet: guess_value [@default `From_metadata];
+  guess_secgroup: guess_value [@default `From_metadata];
+} [@@deriving yojson, make]
+
+module To_genspio = struct
+  open Genspio_edsl
+
+  let saylp t fmt l =
```
Comment (ihodes) [inline src/lib/aws_efs.ml]: What does `lp` stand for?

File: src/lib/aws_efs.ml [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+
+  let wait_for_mount_target_available t ~mount_target_id =
+    seq [
+      saylp t "Waiting for mount-target to be really available." [];
+      seq_succeeds_or
+        ~name:"Waiting-for-mount-target"
+        ~silent:false
+        ~clean_up:[fail] [
+        loop_until_ok
```
Comment (ihodes) [inline src/lib/aws_efs.ml:191]: This is a cool Genspio use!

**Review by ihodes — APPROVED** [review]

File: src/lib/aws_efs.ml (hunk 4 above)
Comment (smondet) [reply src/lib/aws_efs.ml]: `l` for List of `string t` arguments.
`p` for "with `Prompt`

File: src/lib/aws_efs.ml (hunk 2 above)
Comment (smondet) [reply src/lib/aws_efs.ml:35]: Let's say `` `Value "sg-12345" `` is a “very educated guess” :)

Do you have a better name?

**Pull request merged**